Sources/GRPC/ClientConnection.swift

```diff
@@ -132,97 +101,145 @@ public class ClientConnection {
       // We're already shutdown or in the process of shutting down.
       return channel.flatMap { $0.closeFuture }
     } else {
-      self.connectivity.state = .shutdown
+      self.connectivity.initiateUserShutdown()
       return channel.flatMap { $0.close() }
     }
   }
 }
 
+// MARK: - Channel creation
+
 extension ClientConnection {
-  /// Creates a `Channel` using the given configuration.
+  /// Register a callback on the close future of the given `channel` to replace the channel (if
+  /// possible) and also replace the `multiplexer` with that from the new channel.
   ///
-  /// This involves: creating a `ClientBootstrap`, connecting to a target and verifying that the TLS
-  /// handshake was successful (if TLS was configured). We _may_ additiionally set a connection
-  /// timeout and schedule a retry attempt (should the connection fail) if a
-  /// `ConnectionBackoff.Iterator` is provided.
+  /// - Parameter channel: The channel that will be set.
+  private func willSetChannel(to channel: EventLoopFuture<Channel>) {
+    // If we're about to set the channel and the user has initiated a shutdown (i.e. while the new
+    // channel was being created) then it is no longer needed.
+    guard !self.connectivity.userHasInitiatedShutdown else {
+      channel.whenSuccess { channel in
+        channel.close(mode: .all, promise: nil)
+      }
+      return
+    }
+
+    channel.flatMap { $0.closeFuture }.whenComplete { _ in
+      guard self.connectivity.canAttemptReconnect else { return }
+      self.channel = ClientConnection.makeChannel(
+        configuration: self.configuration,
+        connectivity: self.connectivity
+      )
+    }
+
+    self.multiplexer = channel.flatMap {
+      $0.pipeline.handler(type: HTTP2StreamMultiplexer.self)
+    }
+  }
+
+  /// Register a callback on the given `channel` to update the connectivity state.
   ///
-  /// See the individual functions for more information:
-  ///  - `makeBootstrap(configuration:)`, and
-  ///  - `verifyTLS(channel:)`.
+  /// - Parameter channel: The channel that was set.
+  private func didSetChannel(to channel: EventLoopFuture<Channel>) {
```

Comment:
`willSet` is preparing the channel so that it can be used, `didSet` is advertising that `self.channel` may be used (or not in the case of `shutdown`).

If we had the current `didSet` logic in `willSet` then we potentially have race problems when e.g. waiting on `.ready` (since it would be triggered before actually assigning the new value). If the `willSet` logic was in `didSet` then it would be possible to call `self.channel` and hit an error (closing the channel) _before_ the reconnect callback was registered.